File: src/main/java/hes/sudoku/RecursiveSolver.java
package hes.sudoku;

import java.util.*;

public class RecursiveSolver {

    private final Puzzle puzzle;
    private final Printer printer;

    public RecursiveSolver(Puzzle puzzle) {
        this.puzzle = puzzle;
        this.printer = new Printer(puzzle);
    }

    public Set<Move> solve() {
        Set<Move> moves = new LinkedHashSet<>();
        solve(moves);

        return moves;
    }

    public boolean solve(Set<Move> moves) {
        if (puzzle.isSolved()) {
            return true;
        }

        Collection<Move> candidates = getPossibleMoves();

        for (Move candidate : candidates) {
            printer.println("Trying " + candidate);

            Puzzle clonedPuzzle = new Puzzle(puzzle);
            clonedPuzzle.apply(candidate);

            if (solve(clonedPuzzle, moves)) {
                moves.add(candidate);
                return true;
            }
            printer.println("%s does not solve the puzzle", candidate);
        }

        return false;
    }

    private boolean solve(Puzzle puzzle, Set<Move> candidates) {
        LogicSolver solver = new LogicSolver(puzzle, Printer.silentPrinter());
        solver.fillNumbers();
        if (puzzle.isSolved()) {
            printer.println("Puzzle solved recursively.");
            return true;
        }

        new Printer(puzzle).printPuzzle();
        printer.println("Puzzle not solved. Trying recursively.");

        return new RecursiveSolver(puzzle).solve(candidates);
    }

    private Collection<Move> getPossibleMoves() {
        ArrayList<Move> moves = new ArrayList<>();
        for (Cell cell : puzzle.getCells()) {
            for (Integer  number : cell.getCandidates()) {
                moves.add(new Move(cell, number, "Recursive guess out of " + cell.getCandidates().size()));
            }
        }

        moves.sort(new CellCandidateComparator());

        return moves;
    }

    private static class CellCandidateComparator implements Comparator<Move> {
        @Override
        public int compare(Move o1, Move o2) {
            return o1.cell().getCandidates().size() - o2.cell().getCandidates().size();
        }
    }
}

File: src/main/java/hes/sudoku/kt/RecursiveSolver.kt
package hes.sudoku.kt

class RecursiveSolver(private val puzzle: Puzzle) {

    private val printer: Printer = Printer(puzzle)

    fun solve(): Set<Move> {
        val moves = mutableSetOf<Move>()
        solve(moves)

        return moves
    }

    private fun solve(moves: MutableSet<Move>): Boolean {
        if (puzzle.isSolved()) {
            return true
        }
        val candidates = possibleMoves
        for (candidate in candidates) {
            printer.println("Trying $candidate")
            val clonedPuzzle = Puzzle(puzzle)
            clonedPuzzle.apply(candidate)
            if (solve(clonedPuzzle, moves)) {
                moves.add(candidate)
                return true
            }
            printer.println("$candidate does not solve the puzzle")
        }
        return false
    }

    private fun solve(puzzle: Puzzle, candidates: MutableSet<Move>): Boolean {
        val solver = LogicSolver(puzzle, printer.silentPrinter())
        solver.fillNumbers()

        if (puzzle.isSolved()) {
            printer.println("Puzzle solved recursively.")
            return true
        }

        Printer(puzzle).printPuzzle()
        printer.println("Puzzle not solved. Trying recursively.")

        return RecursiveSolver(puzzle).solve(candidates)
    }

    private val possibleMoves: Collection<Move>
        get() {
            val moves = mutableListOf<Move>()
            for (cell in puzzle.cells) {
                for (number in cell.candidates) {
                    moves.add(Move(cell, number).withDescription("Recursive guess out of " + cell.candidates.size))
                }
            }
            moves.sortWith(CellCandidateComparator())

            return moves
        }

    private class CellCandidateComparator : Comparator<Move> {
        override fun compare(o1: Move, o2: Move): Int {
            return o1.cell.candidates.size - o2.cell.candidates.size
        }
    }
}